File: VB/WindowsApplication1/NavBarGroupStateHelper.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.XtraEditors.ViewInfo
Imports DevExpress.XtraEditors.Drawing
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraNavBar
Imports DevExpress.XtraNavBar.ViewInfo
Imports System.Reflection

Namespace WindowsApplication1
	Public Class NavBarGroupStateHelper
		Inherits Component

		Private isLocked As Boolean

		Private _CheckIndent As Integer
		Public Property CheckIndent() As Integer
			Get
				Return _CheckIndent
			End Get
			Set(ByVal value As Integer)
				_CheckIndent = value
			End Set
		End Property

		Private _GroupEdit As RepositoryItemCheckEdit
		<DesignerSerializationVisibility(DesignerSerializationVisibility.Content)>
		Public Property GroupEdit() As RepositoryItemCheckEdit
			Get
				Return _GroupEdit
			End Get
			Set(ByVal value As RepositoryItemCheckEdit)
				_GroupEdit = value
			End Set
		End Property

		Private _NavBarControl As NavBarControl

		Public Property NavBarControl() As NavBarControl
			Get
				Return _NavBarControl
			End Get
			Set(ByVal value As NavBarControl)
				UnsubscribeEvents(value)
				_NavBarControl = value
				SubscribeEvents(value)
			End Set
		End Property

		Public Sub New()
			GroupEdit = New RepositoryItemCheckEdit()
		End Sub


		Private Function IsCustomDrawNeeded() As Boolean
			Return GroupEdit IsNot Nothing AndAlso NavBarControl IsNot Nothing AndAlso Not isLocked
		End Function


		Private Sub SubscribeEvents(ByVal navBarControl As NavBarControl)
			If navBarControl IsNot Nothing Then
				AddHandler Me.NavBarControl.CustomDrawGroupCaption, AddressOf NavBarControl_CustomDrawGroupCaption
				AddHandler Me.NavBarControl.MouseClick, AddressOf NavBarControl_MouseClick
			End If
		End Sub


		Private Sub UnsubscribeEvents(ByVal navBarControl As NavBarControl)
			If navBarControl Is Nothing Then
				RemoveHandler Me.NavBarControl.CustomDrawGroupCaption, AddressOf NavBarControl_CustomDrawGroupCaption
				RemoveHandler Me.NavBarControl.MouseClick, AddressOf NavBarControl_MouseClick
			End If
		End Sub

		Private Function GetCheckBoxWidth() As Integer
			Return CheckIndent * 2 + 10
		End Function

		Private Function GetCaptionBounds(ByVal originalCaptionBounds As Rectangle) As Rectangle
			Return New Rectangle(originalCaptionBounds.Location, New Size(originalCaptionBounds.Width - GetCheckBoxWidth(), originalCaptionBounds.Height))
		End Function

		Private Function GetCheckBoxBounds(ByVal fixedCaptionBounds As Rectangle) As Rectangle
			Return New Rectangle(fixedCaptionBounds.Right, fixedCaptionBounds.Top, GetCheckBoxWidth(), fixedCaptionBounds.Height)

		End Function

		Private Sub NavBarControl_CustomDrawGroupCaption(ByVal sender As Object, ByVal e As DevExpress.XtraNavBar.ViewInfo.CustomDrawNavBarElementEventArgs)
			If Not IsCustomDrawNeeded() Then
				Return
			End If
			Try
				isLocked = True
				Dim painter As BaseNavGroupPainter = NavBarControl.View.CreateGroupPainter(NavBarControl)
				Dim info As NavGroupInfoArgs = TryCast(e.ObjectInfo, NavGroupInfoArgs)

				Dim originalCaptionBounds As New Rectangle(info.CaptionClientBounds.X,info.CaptionClientBounds.Y,info.CaptionClientBounds.Width-info.ButtonBounds.Width,info.CaptionClientBounds.Height)
				Dim fixedCaptionBounds As Rectangle = GetCaptionBounds(originalCaptionBounds)
				Dim checkBoxBounds As Rectangle = GetCheckBoxBounds(fixedCaptionBounds)
				info.CaptionBounds = fixedCaptionBounds
				painter.DrawObject(info)
				info.CaptionBounds = originalCaptionBounds
				DrawCheckBox(e.Graphics, checkBoxBounds, IsGroupEnabled(info.Group))
				e.Handled = True
			Finally
				isLocked = False
			End Try
		End Sub
		Protected Sub DrawCheckBox(ByVal g As Graphics, ByVal r As Rectangle, ByVal Checked As Boolean)
			Dim painter As BaseEditPainter = GroupEdit.CreatePainter()
			Dim info As BaseEditViewInfo = GroupEdit.CreateViewInfo()
			info.EditValue = Checked
			info.Bounds = r
			info.CalcViewInfo(g)
			Dim args As New ControlGraphicsInfoArgs(info, New DevExpress.Utils.Drawing.GraphicsCache(g), r)
			painter.Draw(args)
			args.Cache.Dispose()
		End Sub
		Private hotRectangle As Rectangle
		Private hotGroup As NavBarGroup

		Private Shared Function GetNavBarView(ByVal NavBar As NavBarControl) As NavBarViewInfo
			Dim pi As PropertyInfo = GetType(NavBarControl).GetProperty("ViewInfo", BindingFlags.Instance Or BindingFlags.NonPublic)
			Return TryCast(pi.GetValue(NavBar, Nothing), NavBarViewInfo)
		End Function

		Private Function IsCheckBox(ByVal p As Point) As Boolean
			Dim hi As NavBarHitInfo = NavBarControl.CalcHitInfo(p)
			If hi.Group Is Nothing Then
				Return False
			End If

			Dim vi As NavBarViewInfo = GetNavBarView(NavBarControl)
			vi.Calc(NavBarControl.ClientRectangle)
			Dim groupInfo As NavGroupInfoArgs = vi.GetGroupInfo(hi.Group)
			Dim originalCaptionBounds As New Rectangle(groupInfo.CaptionClientBounds.X, groupInfo.CaptionClientBounds.Y, groupInfo.CaptionClientBounds.Width - groupInfo.ButtonBounds.Width, groupInfo.CaptionClientBounds.Height)
			Dim checkBounds As Rectangle = GetCheckBoxBounds(GetCaptionBounds(originalCaptionBounds))
			hotGroup = hi.Group
			hotRectangle = checkBounds
			Return checkBounds.Contains(p)
		End Function

		Private Sub NavBarControl_MouseClick(ByVal sender As Object, ByVal e As MouseEventArgs)
			If IsCheckBox(e.Location) Then
				ToggleGroupState(hotGroup)
			End If
		End Sub

		Private Shared Function IsGroupEnabled(ByVal group As NavBarGroup) As Boolean
			Return group.Tag Is Nothing
		End Function

		Private Sub ToggleGroupState(ByVal group As NavBarGroup)
			SetGroupState(group, (Not IsGroupEnabled(group)))
		End Sub

		Private Sub SetGroupState(ByVal group As NavBarGroup, ByVal enabled As Boolean)
			For Each link As NavBarItemLink In group.ItemLinks
				link.Item.Enabled = enabled
			Next link
			If enabled Then
				group.Tag = Nothing
			Else
				group.Tag = 0
			End If
			NavBarControl.Invalidate(hotRectangle)
		End Sub
	End Class
End Namespace

File: CS/WindowsApplication1/NavBarGroupStateHelper.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraEditors.ViewInfo;
using DevExpress.XtraEditors.Drawing;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraNavBar;
using DevExpress.XtraNavBar.ViewInfo;
using System.Reflection;

namespace WindowsApplication1
{
    public class NavBarGroupStateHelper : Component
    {
        private bool isLocked;

        private int _CheckIndent;
        public int CheckIndent
        {
            get { return _CheckIndent; }
            set { _CheckIndent = value; }
        }

        private RepositoryItemCheckEdit _GroupEdit;
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
        public RepositoryItemCheckEdit GroupEdit
        {
            get { return _GroupEdit; }
            set { _GroupEdit = value; }
        }

        private NavBarControl _NavBarControl;
                                     
        public NavBarControl NavBarControl
        {
            get { return _NavBarControl; }
            set { UnsubscribeEvents(value); _NavBarControl = value; SubscribeEvents(value); }
        }

        public NavBarGroupStateHelper()
        {
            GroupEdit = new RepositoryItemCheckEdit();
        }
             

        bool IsCustomDrawNeeded()
        {
            return GroupEdit != null && NavBarControl != null && !isLocked;
        }


        private void SubscribeEvents(NavBarControl navBarControl)
        {
            if (navBarControl != null)
            {
                NavBarControl.CustomDrawGroupCaption += NavBarControl_CustomDrawGroupCaption;
                NavBarControl.MouseClick += NavBarControl_MouseClick;
            }
        }

     
        private void UnsubscribeEvents(NavBarControl navBarControl)
        {
            if (navBarControl == null)
            {
                NavBarControl.CustomDrawGroupCaption -= NavBarControl_CustomDrawGroupCaption;
                NavBarControl.MouseClick -= NavBarControl_MouseClick;
            }
        }

        int GetCheckBoxWidth()
        {
            return CheckIndent * 2 + 10;
        }

        Rectangle GetCaptionBounds(Rectangle originalCaptionBounds)
        {
            return new Rectangle(originalCaptionBounds.Location, new Size(originalCaptionBounds.Width - GetCheckBoxWidth(), originalCaptionBounds.Height));
        }

        Rectangle GetCheckBoxBounds(Rectangle fixedCaptionBounds)
        {
            return new Rectangle(fixedCaptionBounds.Right, fixedCaptionBounds.Top, GetCheckBoxWidth(), fixedCaptionBounds.Height); ;
        }

        void NavBarControl_CustomDrawGroupCaption(object sender, DevExpress.XtraNavBar.ViewInfo.CustomDrawNavBarElementEventArgs e)
        {
            if (!IsCustomDrawNeeded()) return;
            try
            {
                isLocked = true;
                BaseNavGroupPainter painter = NavBarControl.View.CreateGroupPainter(NavBarControl);
                NavGroupInfoArgs info = e.ObjectInfo as NavGroupInfoArgs; 
               
                Rectangle originalCaptionBounds = new Rectangle(info.CaptionClientBounds.X,info.CaptionClientBounds.Y,info.CaptionClientBounds.Width-info.ButtonBounds.Width,info.CaptionClientBounds.Height);
                Rectangle fixedCaptionBounds = GetCaptionBounds(originalCaptionBounds);
                Rectangle checkBoxBounds = GetCheckBoxBounds(fixedCaptionBounds);
                info.CaptionBounds = fixedCaptionBounds;
                painter.DrawObject(info);
                info.CaptionBounds = originalCaptionBounds;
                DrawCheckBox(e.Graphics, checkBoxBounds, IsGroupEnabled(info.Group));
                e.Handled = true;
            }
            finally
            {
                isLocked = false;
            }
        }
        protected void DrawCheckBox(Graphics g, Rectangle r, bool Checked)
        {
            BaseEditPainter painter = GroupEdit.CreatePainter();
            BaseEditViewInfo info = GroupEdit.CreateViewInfo();
            info.EditValue = Checked;
            info.Bounds = r;
            info.CalcViewInfo(g);
            ControlGraphicsInfoArgs args = new ControlGraphicsInfoArgs(info, new DevExpress.Utils.Drawing.GraphicsCache(g), r);
            painter.Draw(args);
            args.Cache.Dispose();
        }
        private Rectangle hotRectangle;
        private NavBarGroup hotGroup;

        static NavBarViewInfo GetNavBarView(NavBarControl NavBar)
        {
            PropertyInfo pi = typeof(NavBarControl).GetProperty("ViewInfo", BindingFlags.Instance | BindingFlags.NonPublic);
            return pi.GetValue(NavBar, null) as NavBarViewInfo;
        }

        bool IsCheckBox(Point p)
        {
            NavBarHitInfo hi = NavBarControl.CalcHitInfo(p);
            if (hi.Group == null) return false; ;
            NavBarViewInfo vi = GetNavBarView(NavBarControl);
            vi.Calc(NavBarControl.ClientRectangle);
            NavGroupInfoArgs groupInfo = vi.GetGroupInfo(hi.Group);
            Rectangle originalCaptionBounds = new Rectangle(groupInfo.CaptionClientBounds.X, groupInfo.CaptionClientBounds.Y, groupInfo.CaptionClientBounds.Width - groupInfo.ButtonBounds.Width, groupInfo.CaptionClientBounds.Height);
            Rectangle checkBounds = GetCheckBoxBounds(GetCaptionBounds(originalCaptionBounds));
            hotGroup = hi.Group;
            hotRectangle = checkBounds;
            return checkBounds.Contains(p);
        }

        void NavBarControl_MouseClick(object sender, MouseEventArgs e)
        {
            if (IsCheckBox(e.Location)) ToggleGroupState(hotGroup);
        }

        static bool IsGroupEnabled(NavBarGroup group)
        {
            return group.Tag == null;
        }

        void ToggleGroupState(NavBarGroup group)
        {
            SetGroupState(group, !IsGroupEnabled(group));
        }

        void SetGroupState(NavBarGroup group, bool enabled)
        {
            foreach (NavBarItemLink link in group.ItemLinks)
            {
                link.Item.Enabled = enabled;
            }
            if (enabled) group.Tag = null;
            else group.Tag = 0;
            NavBarControl.Invalidate(hotRectangle);
        }
    }
}
